packit/dg_robot.py

```diff
@@ -0,0 +1,256 @@
+        """
+        f = FedPKG(self.fas_user, self.distgit.working_dir)
+        f.init_ticket()
+        f.new_sources(sources=archive_path)
+
+    def sync_files(self, files_to_sync: List[str]) -> None:
+        """
+        sync required files from upstream to downstream
+        """
+        for fi in files_to_sync:
+            fi = fi[1:] if fi.startswith("/") else fi
+            src = os.path.join(self.upstream.working_dir, fi)
+            shutil.copy2(src, self.distgit.working_dir)
+
+    def create_branch_distgit(self, branch_nane: str):
+        # fetch and reset --hard master?
+        # what if the branch already exists?
+        self.distgit.git_repo.create_head(branch_nane)
+
+    def checkout_branch_distgit(self, git_ref: str):
+        self.distgit.git_repo.heads[git_ref].checkout()
+
+    def commit_distgit(self, title: str, msg: str) -> None:
+        main_msg = f"[packit] {title}"
+        self.distgit.git_repo.git.add("-A")
+        self.distgit.git_repo.index.write()
+        # TODO: attach git note to every commit created
+        # TODO: implement cleaning policy: once the PR is closed (merged/refused), remove the branch
+        #       make this configurable so that people know this would happen, don't clean by default
+        #       we should likely clean only merged PRs by default
+        # TODO: implement signing properly: we need to create a cert for the bot, distribute it to the container,
+        #       prepare git config and then we can start signing
+        # TODO: make -s configurable
+        self.distgit.git_repo.git.commit("-s", "-m", main_msg, "-m", msg)
+
+    def push_distgit_fork(self, branch_name: str, fork_remote_name: str = "fork"):
+        """
+        push changes to the dist-git repo; they need to be committed!
+
+        :param branch_name: the branch where we push
+        :param fork_remote_name: local name of the remote where we push to
+        """
+        if fork_remote_name not in [
+            remote.name for remote in self.distgit.git_repo.remotes
+        ]:
+            fork_urls = self.distgit.git_project.get_fork().get_git_urls()
+            self.distgit.git_repo.create_remote(
+                name=fork_remote_name, url=fork_urls["ssh"]
+            )
+
+        # I suggest to comment this one while testing when the push is not needed
+        # TODO: create dry-run ^
+        fork_branches = self.distgit.git_project.get_fork().get_branches()
+        # self.distgit.git_repo.remote(fork_remote_name).push(
+        #     refspec=branch_name, force=branch_name in fork_branches
+        # )
+
+    def create_pull(
+            self,
+            pr_title: str,
+            pr_description: str,
+            source_branch: str,
+            target_branch: str,
+    ) -> None:
+        """
+        use the content of the dist-git repo and create a PR
+        :return:
+        """
+        import ipdb; ipdb.set_trace()
+        # FIXME: project is None
```

Comment:
I can take a look.

If you set `git_service` and `full_name` (or `repo_name` + `namespace`) in LocalProject, it should create the `git_project`.

Reply:
I figured it out in the end.